java-checks/src/main/java/org/sonar/java/checks/synchronization/TwoLocksWaitNotifyCheck.java

```diff
@@ -0,0 +1,176 @@
+      return;
+    }
+    if (tree.is(METHOD) || tree.is(CONSTRUCTOR)) {
+      synchronizedStack.pop();
+    }
+  }
+
+  private void methodTree(Tree tree) {
+    findMethodCall(tree, WAIT_MATCHER)
+      .ifPresent(wait -> reportIssue(wait, "Don't use \"wait()\" here; multiple locks are held.", flowFromTree(tree), null));
+    findMethodCall(tree, NOTIFY_MATCHER)
+      .ifPresent(notify -> reportIssue(notify, "Use \"notifyAll()\" here; multiple locks are held.", flowFromTree(tree), null));
+  }
+
+  private Optional<MethodInvocationTree> findMethodCall(Tree tree, MethodMatcher methodMatcher) {
+    MethodInvocationVisitor visitor = new MethodInvocationVisitor(methodMatcher);
+    tree.accept(visitor);
+    return visitor.matchedMethods().findAny();
+  }
+
+  private static List<JavaFileScannerContext.Location> flowFromTree(Tree tree) {
+    SynchronizedKeywordVisitor synchronizedKeywordVisitor = new SynchronizedKeywordVisitor();
+    tree.accept(synchronizedKeywordVisitor);
+    return synchronizedKeywordVisitor.stream()
+      .map(t -> new JavaFileScannerContext.Location("locking", t))
+      .collect(Collectors.toList());
+  }
+
+  private class MethodInvocationVisitor extends BaseTreeVisitor {
+
+    private final MethodMatcher methodMatcher;
+    private Stream.Builder<MethodInvocationTree> matchedMethods = Stream.builder();
+
+    private MethodInvocationVisitor(MethodMatcher methodMatcher) {
+      this.methodMatcher = methodMatcher;
+    }
+
+    @Override
+    public void visitSynchronizedStatement(SynchronizedStatementTree tree) {
+      synchronizedStack.peek().inc();
+      super.visitSynchronizedStatement(tree);
+      synchronizedStack.peek().dec();
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
+      if (methodMatcher.matches(tree) && synchronizedStack.peek().value >= 2) {
+        matchedMethods.add(tree);
+      }
+    }
+
+    @Override
+    public void visitClass(ClassTree tree) {
+      // cut on anonymous and local classes
+    }
+
+    private Stream<MethodInvocationTree> matchedMethods() {
+      return matchedMethods.build();
+    }
+  }
+
+  private static class SynchronizedKeywordVisitor extends BaseTreeVisitor {
+
+    private Stream.Builder<SyntaxToken> synchronizedKeywords = Stream.builder();
+
+    @Override
+    public void visitSynchronizedStatement(SynchronizedStatementTree tree) {
+      synchronizedKeywords.add(tree.synchronizedKeyword());
+      super.visitSynchronizedStatement(tree);
+    }
+
+    @Override
+    public void visitMethod(MethodTree tree) {
+      findModifier(tree.modifiers(), SYNCHRONIZED)
+        .ifPresent(s -> synchronizedKeywords.add(s.keyword()));
+      super.visitMethod(tree);
+    }
+
+    Stream<SyntaxToken> stream() {
+      return synchronizedKeywords.build();
+    }
+  }
+
+  private static class Counter {
+    int value;
+
+    private Counter(int initialValue) {
+      this.value = initialValue;
+    }
+
+    int inc() {
```

Comment:
matter of personal preference but I would go for full name rather than abbreviation. Do as you think is better.